### PR TITLE
Character Joint tool

### PR DESCRIPTION
**Pull request by Wesley201**
# Peer Review Information:
Separated the character tool from validation and added to perception 

Creates the nose, left and right ear joints and adds the joint label script to the game objects. The game objects use the base coco template when the label is added, also the joints are parented to the head 

I need to create tests for the character tool only, left the validation tests in the separate suite 

## Editor / Package versioning:
**Editor Version Target**: 2019.4

## Dev Testing:
**Tests Added**: 

**Core Scenario Tested**: 
Tested the tool using the up to date layouts of the render people in the content db 

**At Risk Areas**: 

## Checklist
- [ ] - Updated docs
- [x] - Updated changelog
- [ ] - Updated test rail

### CODE REVIEW

**Comment by mkamalza** [comment]
Thanks Wes. I haven't had time to review the code yet, but here is my initial feedback on the UI. I think we will eventually have to redesign the window into a new perception asset tools window, but for now, we need to address most, if not all of these:
- If no object is selected or a the selected object is not the root of a rigged humanoid, the window should inform the user and disable all buttons. Currently, if you try the buttons on an unsupported object, there are exceptions, but no feedback in the UI.
- Tabs should probably be named Keypoints and Validation? Since the whole window is for Character tools. Not sure about the exact names though.
- The window title should match the title used in the menu option
- I think renaming the window and menu option to "Perception Character Tool" would be better, to make it stand out more.
- We should have tooltips on each UI element to explain what it means. E.g. what does validate bones do, or what does "Character is Valid" mean.
- If the character already has nose and ear joints, the button should be disabled, possibly with a little UI text saying the character already has them.
- I'm not sure what "Draw Face Rays" does. It doesn't seem to do anything when I enable it.
- When I click "Validate Bones", nothing seems to happen. There should be some kind of UI feedback. Same with "Validate Pose Data".
- In general, in case of both successful or unsuccessful operations, there should be a message on the UI that communicates this to the user. This could be a label at the bottom of the window for now.

File: com.unity.perception/Editor/Character/ModelTestUI.cs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+                                    }
+                                }
+                            }
+                        }
+                        else if (failedBones.Count == 0)
+                        {
+                            GUILayout.Label(string.Format("Required Bones Present : {0}", TestResults.Pass), EditorStyles.whiteLabel);
+                        }
+
+                        if (test)
```
Comment (StevenBorkman) [inline com.unity.perception/Editor/Character/ModelTestUI.cs]: this could be reduced to: testResults = test ? TestResults.Pass : TestResults.Fail;

File: com.unity.perception/Editor/Character/ModelTestUI.cs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                            testResults = TestResults.Fail;
+                    }
+
+                    if (GUILayout.Button("Validate Pose Data", GUILayout.Width(160)))
+                    {
+                        testResults = TestResults.Running;
+
+                        var test = contentTests.CharacterPoseData(selection, out failedPose);
+
+                        if (test)
```
Comment (StevenBorkman) [inline com.unity.perception/Editor/Character/ModelTestUI.cs]: See previous comment

File: com.unity.perception/Tests/Editor/CharacterToolingTests.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            selectionLists = new List<GameObject>();
+            AssetListCreation();
+        }
+
+        [TearDown]
+        public void TearDown()
+        {
+            selectionLists.Clear();
+        }
+
```
Comment (StevenBorkman) [inline com.unity.perception/Tests/Editor/CharacterToolingTests.cs:29]: I think you are going to need tests where creating ears & nose fails (if that is possible to do) also tests for the validation, perhaps test against a valid model and also at least one test where the model does not pass validation

**Review by StevenBorkman — CHANGES_REQUESTED** [review]
Definitely think we need better comments & tests. Some of the math was hard to follow, and I'll re-review after it has been commented so I can tell what is going on.

File: com.unity.perception/Editor/Validation/CharacterTooling.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        public bool CharacterPoseData(GameObject gameObject, out List<GameObject> failedGameObjects)
+        {
+            failedGameObjects = new List<GameObject>();
+
+            var componentsParent = gameObject.GetComponents<Transform>();
+            var componentsChild = gameObject.GetComponentsInChildren<Transform>();
+
+            for (int p = 0; p < componentsParent.Length; p++)
+            {
+                if (componentsParent[p].GetType() == typeof(Transform))
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Validation/CharacterTooling.cs]: not worried about the Get Components not working it is more of the fbx being imported something becoming corrupted, like a backwards update issue or something like a change in the prefab code that causes older imported fbx to no have null values or errors.

File: com.unity.perception/Editor/Validation/CharacterTooling.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                return true;
+
+            return false;
+        }
+
+        public bool CharacterPoseData(GameObject gameObject, out List<GameObject> failedGameObjects)
+        {
+            failedGameObjects = new List<GameObject>();
+
+            var componentsParent = gameObject.GetComponents<Transform>();
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Validation/CharacterTooling.cs]: True, it was easier for me to read to make sure I grabbed the right things from the parent and the children because of the large number of children. I can always change it

File: com.unity.perception/Editor/Validation/CharacterTooling.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                var boneValue = bone.Value;
+
+                if (boneValue != true)
+                    failed.Add(boneKey, boneValue);
+            }
+
+            if (failed.Count == 0)
+                return true;
+
+            return false;
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Validation/CharacterTooling.cs]: fixed

File: com.unity.perception/Editor/Validation/CharacterTooling.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
+using System.Collections.Generic;
+using static UnityEngine.Perception.Content.CharacterValidation;
+using System.Linq;
+
+namespace UnityEngine.Perception.Content
+{
+    public class CharacterTooling : MonoBehaviour
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Validation/CharacterTooling.cs]: I added some summary comments in the scripts, let me know if that is ok for now?

File: com.unity.perception/Tests/Editor/CharacterToolingTests.cs (hunk 3 above)
Comment (Wesley201) [reply com.unity.perception/Tests/Editor/CharacterToolingTests.cs:29]: Plan is next task will have tests added for character tool in that PR, I did add a few tests in this PR as well

File: com.unity.perception/Editor/Validation/CharacterTooling.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Validation/CharacterTooling.cs]: This is needed for fbx objects if they are selected

File: com.unity.perception/Editor/Validation/CharacterTooling.cs [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+            var componentsChild = gameObject.GetComponentsInChildren<Transform>();
+
+            for (int p = 0; p < componentsParent.Length; p++)
+            {
+                if (componentsParent[p].GetType() == typeof(Transform))
+                {
+                    var pos = componentsParent[p].transform.position;
+                    var rot = componentsParent[p].transform.rotation.eulerAngles;
+
+                    if (pos == null || rot == null)
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Validation/CharacterTooling.cs]: I have a check in here mainly because I found user cases where users would input invalid fbx files and there would be a messed up transform with null values

File: com.unity.perception/Editor/Validation/CharacterTooling.cs [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+            {
+                if (componentsChild[c].GetType() == typeof(Transform))
+                {
+                    var pos = componentsChild[c].transform.position;
+                    var rot = componentsChild[c].transform.rotation.eulerAngles;
+
+                    if (pos == null || rot == null)
+                    {
+                        failedGameObjects.Add(componentsChild[c].gameObject);
+                    }
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Validation/CharacterTooling.cs]: above

File: com.unity.perception/Editor/Character/CharacterTooling.cs [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+                if (boneValue != true)
+                    failed.Add(boneKey, boneValue);
+            }
+
+            return failed.Count == 0;
+        }
+
+        /// <summary>
+        /// Ensures there is pose data in the parent and child game objects of a character by checking for pos and rot
```
Comment (StevenBorkman) [inline com.unity.perception/Editor/Character/CharacterTooling.cs:35]: I would go ahead and write out 'position' and 'rotation' since this API comments

File: com.unity.perception/Editor/Character/CharacterToolingLibrary.cs [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+            Debug.DrawRay(earCenter, Vector3.right, Color.blue, duration);
+            Debug.DrawRay(earCenter, Vector3.left, Color.green, duration);
+        }
+
+        public static GameObject CreateNewCharacterPrefab(GameObject selection, Vector3 nosePosition, Vector3 earRightPosition, Vector3 earLeftPosition)
+        {
+            var root = (GameObject)PrefabUtility.InstantiatePrefab(selection);
+            if(root == null)
+                root = Object.Instantiate(selection);
+
```
[the remark this answers is not in the document]
Comment (StevenBorkman) [reply com.unity.perception/Editor/Character/CharacterToolingLibrary.cs:322]: I still think this issue exists.

**Review by StevenBorkman — CHANGES_REQUESTED** [review]
Still some things to resolve, and I don't know if I'm missing something but some of the things marked resolved don't really appear to be resolved, or even commented on why they shouldn't be resolved, so I re-opened them.

File: com.unity.perception/Editor/Character/CharacterTooling.cs (hunk 10 above)
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterTooling.cs:35]: Fixed in next changeset, will push later today

File: com.unity.perception/Editor/Character/CharacterToolingLibrary.cs [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+            return model;
+        }
+
+        private static void AddJointLabel(GameObject gameObject)
+        {
+            var jointLabel = gameObject.AddComponent<JointLabel>();
+            var data = new JointLabel.TemplateData();
+            var asset = new Object();
+            var template = AssetDatabase.GetAllAssetPaths().Where(o => o.EndsWith("CocoKeypointTemplate.asset", StringComparison.OrdinalIgnoreCase)).ToList();
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterToolingLibrary.cs:370]: Interesting I was under the impression that template was leaving the joint label script. Has this changed?

File: com.unity.perception/Editor/Character/CharacterToolingLibrary.cs [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+                var distanceY = Math.Abs(pointR.y - pointL.y);
+
+                if (distanceX < 0.01 && distanceY < 0.01 )
+                {
+                    faceCenter = pointR;
+                }
+            }
+            
+            rayHead.origin = head.position;
+            rayHead.direction = Vector3.up * distanceCheck;
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterToolingLibrary.cs:170]: Yes, I was was giving the ray a certain distance I wanted it to dray so I could check the point on the ray out to a certain distance. 

More or less, dray the ray up on the Vector 3 at this distance

File: com.unity.perception/Editor/Character/CharacterToolingLibrary.cs [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+            }
+            else
+            {
+                eyeDistance = Vector3.Distance(leftEye.position, rightEye.position);
+                directionLeft = Quaternion.AngleAxis(-45, -leftEye.right) * -leftEye.up;
+                directionRight = Quaternion.AngleAxis(-45, rightEye.right) * -rightEye.up;
+            }
+
+            rayLeftEye.origin = leftEye.position;
+            rayLeftEye.direction = directionLeft * eyeDistance;
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterToolingLibrary.cs:149]: I don't understand what you are pointing out here?

File: com.unity.perception/Editor/Character/CharacterToolingLibrary.cs [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+
+            if(leftEye == null || rightEye == null)
+            {
+                Debug.LogWarning("Eye positions are null, unable to position nose joint!");
+            }
+            else
+            {
+                eyeDistance = Vector3.Distance(leftEye.position, rightEye.position);
+                directionLeft = Quaternion.AngleAxis(-45, -leftEye.right) * -leftEye.up;
+                directionRight = Quaternion.AngleAxis(-45, rightEye.right) * -rightEye.up;
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterToolingLibrary.cs:145]: for some reason I'm not understanding this one

File: com.unity.perception/Editor/Character/CharacterToolingLibrary.cs [hunk 16]
```diff
@@ -0,0 +1,331 @@
...
+                var pointL = rayLeftEye.GetPoint(point);
+
+                var distanceX = Math.Abs(pointR.x - pointL.x);
+                var distanceY = Math.Abs(pointR.y - pointL.y);
+
+                if (distanceX < 0.01 && distanceY < 0.01 )
+                {
+                    faceCenter = pointR;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterToolingLibrary.cs]: No I don't think depending on the angle of the eyes and where they collide

File: com.unity.perception/Editor/Character/CharacterToolingLibrary.cs [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+                var pointF = rayNoseBack.GetPoint(point);
+
+                var distanceZ = Math.Abs(pointH.z - pointF.z);
+
+                if (distanceZ < 0.01)
+                {
+                    earCenter = pointF;
+                }
+            }
+
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterToolingLibrary.cs:191]: I will start adding comments

File: com.unity.perception/Editor/Validation/CharacterTooling.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (StevenBorkman) [reply com.unity.perception/Editor/Validation/CharacterTooling.cs]: fair enough

File: com.unity.perception/Editor/Character/CharacterToolingLibrary.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterToolingLibrary.cs:372]: Changed it so there is now a field to select a template and if it the there is no selection, the default is the coco template

File: com.unity.perception/Editor/Character/ModelTestUI.cs [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+    private static string[] toolbarNames = null;
+    private enum TestResults
+    {
+        Inconclusive,
+        Pass,
+        Fail,
+        Running
+    }
+
+    private TestResults testResults = new TestResults();
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/ModelTestUI.cs]: fixed

File: com.unity.perception/Editor/Character/ModelTestUI.cs [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+using UnityEditor;
+using UnityEngine.Perception.Content;
+using MenuItem = UnityEditor.MenuItem;
+using UnityEngine;
+using static UnityEngine.Perception.Content.CharacterValidation;
+using System.Linq;
+
+public class AssetValidation : EditorWindow
+{
+    private static string[] toolbarNames = null;
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/ModelTestUI.cs]: fixed

File: com.unity.perception/Editor/Validation/CharacterTooling.cs [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+            if (failedGameObjects.Count == 0)
+                return true;
+
+            return false;
+        }
+
+        public bool CharacterCreateNose(GameObject selection, Animator animator, SkinnedMeshRenderer skinnedMeshRenderer, bool drawRays = false)
+        {
+            var model = AvatarCreateNose(selection, animator, skinnedMeshRenderer, drawRays);
+            var childCount = model.transform.childCount;
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Validation/CharacterTooling.cs]: Separated the code into 2 separate API's

File: com.unity.perception/Editor/Validation/CharacterTooling.cs [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+            var nose = false;
+            var earRight = false;
+            var earLeft = false;
+
+            for (int i = 0; i < childCount; i++)
+            {
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Validation/CharacterTooling.cs]: I think that was fixed in one of the change sets

File: com.unity.perception/Editor/Character/ModelTestUI.cs [hunk 22]
```diff
@@ -0,0 +1,160 @@
+using System.Collections.Generic;
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/ModelTestUI.cs]: Changed the name

File: com.unity.perception/Editor/Character/CharacterToolingLibrary.cs [hunk 23]
```diff
@@ -0,0 +1,331 @@
...
+                    }
+                }
+            }
+
+            var model = PrefabUtility.SaveAsPrefabAsset(root, "Assets/" + root.name + ".prefab");
+
+            return model;
+        }
+
+        private static void AddJointLabel(GameObject gameObject)
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterToolingLibrary.cs]: removed the private modifier

File: com.unity.perception/Editor/Character/CharacterToolingLibrary.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterToolingLibrary.cs:304]: Pretty sure I fixed this

**Review by StevenBorkman — APPROVED** [review]

File: com.unity.perception/Editor/Character/CharacterToolingUI.cs [hunk 24]
```diff
@@ -0,0 +1,159 @@
...
+        Running
+    }
+
+    TestResults m_testResults = new TestResults();
+    CharacterTooling m_contentTests = new CharacterTooling();
+
+    GameObject selection = null;
+    int toolbarSelection = 0;
+    bool drawFaceRays = false;
+    string savePath = "Assets/";
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterToolingUI.cs]: fixed

File: com.unity.perception/Editor/Character/CharacterToolingUI.cs [hunk 25]
```diff
@@ -0,0 +1,159 @@
...
+using UnityEngine.Perception.Content;
+using MenuItem = UnityEditor.MenuItem;
+using UnityEngine;
+using static UnityEngine.Perception.Content.CharacterValidation;
+using System.Linq;
+
+public class CharacterToolingUI : EditorWindow
+{
+    private static string[] toolbarNames = null;
+    private enum TestResults
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterToolingUI.cs]: fixed

File: com.unity.perception/Editor/Character/CharacterToolingUI.cs [hunk 26]
```diff
@@ -0,0 +1,159 @@
...
+using UnityEditor;
+using UnityEngine.Perception.Content;
+using MenuItem = UnityEditor.MenuItem;
+using UnityEngine;
+using static UnityEngine.Perception.Content.CharacterValidation;
+using System.Linq;
+
+public class CharacterToolingUI : EditorWindow
+{
+    private static string[] toolbarNames = null;
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterToolingUI.cs]: fixed

File: com.unity.perception/Editor/Character/CharacterTooling.cs [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+            {
+                if (jointLabels[i].name.Contains("nose"))
+                    nose = true;
+                if (jointLabels[i].name.Contains("earRight"))
+                    earRight = true;
+                if (jointLabels[i].name.Contains("earLeft"))
+                    earLeft = true;
+            }
+
+            if (nose && earRight && earLeft)
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterTooling.cs:108]: fixed

File: com.unity.perception/Editor/Character/CharacterTooling.cs [hunk 28]
```diff
@@ -0,0 +1,113 @@
...
+        /// <summary>
+        /// Bool function to make create a new prefab Character with nose and ear joints
+        /// </summary>
+        /// <param name="selection"></param>
+        /// <param name="drawRays"></param>
+        /// <param name="savePath"></param>
+        /// <returns></returns>
+        public bool CharacterCreateNose(GameObject selection, bool drawRays = false, string savePath = "Assets/")
+        {
+            var model = AvatarCreateNoseEars(selection, savePath, drawRays);
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterTooling.cs]: fixed

File: com.unity.perception/Editor/Character/CharacterTooling.cs [hunk 29]
```diff
@@ -0,0 +1,113 @@
+using System.Collections.Generic;
+using static UnityEngine.Perception.Content.CharacterValidation;
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterTooling.cs:2]: fixed

File: com.unity.perception/Editor/Character/CharacterTooling.cs [hunk 30]
```diff
@@ -0,0 +1,113 @@
...
+namespace UnityEngine.Perception.Content
+{
+    public class CharacterTooling : MonoBehaviour
+    {
+        /// <summary>
+        /// Bool function used for testing to make sure the target character has the required 15 starting bones
+        /// </summary>
+        /// <param name="selection">target character selected</param>
+        /// <param name="failed">Dictionary return if of Human Bones that tracks they are prsent or missing</param>
+        /// <returns></returns>
```
[the remark this answers is not in the document]
Comment (Wesley201) [reply com.unity.perception/Editor/Character/CharacterTooling.cs:15]: fixed
